# Sami Imran

<h1> First steps:</h1>

1. Import pandas library
2. Read csv file and store it in variable 'df'
3. Retrive random data(sample) and store it
4. Display piece of data

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('online_retail_II.csv')

In [3]:
df = df.sample(200000)

In [4]:
df.head(10)

Invoice StockCode                         Description  Quantity  \
987378  575955     22337      ANGEL DECORATION PAINTED ZINC          6   
900767  569457     22302            COFFEE MUG PEARS  DESIGN         4   
524055  538071     22767         TRIPLE PHOTO FRAME CORNICE          1   
300998  518681     21928      JUMBO BAG SCANDINAVIAN PAISLEY         4   
931091  571716     21504                SKULLS GREETING CARD        10   
940372  572499     23373    VINTAGE CHRISTMAS PAPER GIFT BAG        10   
696500  551351     21442          GREEN BIRDHOUSE DECORATION        24   
545200  537888     22951     60 CAKE CASES DOLLY GIRL DESIGN        24   
752129  556812     20717              STRAWBERRY SHOPPER BAG         1   
432033  530662     22488  NATURAL SLATE RECTANGLE CHALKBOARD        12   

                InvoiceDate  Price  Customer ID         Country  
987378  2011-11-13 12:05:00   0.65      14044.0  United Kingdom  
900767  2011-10-04 11:29:00   2.55      14606.0  United Kingdom  
524055  2010-12-09 14:09:00  20.38          NaN  United Kingdom  
300998  2010-08-10 17:09:00   1.95      17799.0  United Kingdom  
931091  2011-10-18 17:05:00   0.83          NaN  United Kingdom  
940372  2011-10-24 14:10:00   0.82      15786.0  United Kingdom  
696500  2011-04-28 10:20:00   0.72      14298.0  United Kingdom  
545200  2010-12-09 10:04:00   0.55      15358.0  United Kingdom  
752129  2011-06-14 17:25:00   2.46          NaN  United Kingdom  
432033  2010-11-04 10:42:00   1.65      15255.0  United Kingdom

In [5]:
df.tail(10)

Invoice StockCode                         Description  Quantity  \
180448    506463     47566                       PARTY BUNTING        15   
445506   C531807     21355             TOAST ITS - I LOVE YOU        -12   
897291    569228    84031A  CHARLIE+LOLA RED HOT WATER BOTTLE          1   
737012    555383     23182      TOILET SIGN OCCUPIED OR VACANT         2   
398723    527492     22733   3D TRADITIONAL CHRISTMAS STICKERS        12   
461305    533003    18098C      PORCELAIN BUTTERFLY OIL BURNER         3   
894847    569027     23549                   WRAP BIRD GARDEN         25   
850666    565432    84032A  CHARLIE+LOLA PINK HOT WATER BOTTLE         1   
946243    572909     23196       VINTAGE LEAF MAGNETIC NOTEPAD         4   
1012365   577754     22819           BIRTHDAY CARD, RETRO SPOT        36   

                 InvoiceDate  Price  Customer ID         Country  
180448   2010-04-30 07:58:00   4.65      16258.0  United Kingdom  
445506   2010-11-09 16:12:00   1.25      15215.0  United Kingdom  
897291   2011-10-02 14:47:00   4.65      15547.0  United Kingdom  
737012   2011-06-02 15:13:00   0.83      12517.0         Germany  
398723   2010-10-18 11:57:00   1.25      16395.0  United Kingdom  
461305   2010-11-15 15:03:00   2.95      16399.0  United Kingdom  
894847   2011-09-30 10:19:00   0.42      13089.0  United Kingdom  
850666   2011-09-04 13:16:00   2.95      16241.0  United Kingdom  
946243   2011-10-26 15:48:00   1.45      15821.0  United Kingdom  
1012365  2011-11-21 14:47:00   0.42      18139.0  United Kingdom

In [6]:
# snake_case = {'stockcode': 'stock_code', 'invoicedate': 'invoice_date', 'customer id': 'customer_id'}
# df.columns = map(str.lower, df.columns)
# df = df.rename(columns=snake_case)
# df

## Get to know to the dataframe

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 987378 to 1012365
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      200000 non-null  object 
 1   StockCode    200000 non-null  object 
 2   Description  199188 non-null  object 
 3   Quantity     200000 non-null  int64  
 4   InvoiceDate  200000 non-null  object 
 5   Price        200000 non-null  float64
 6   Customer ID  154629 non-null  float64
 7   Country      200000 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 13.7+ MB


## Metadata:
- **We have 8 columns**
- **2 columns have missing data**

## Descriptive statistics
**we care only about quantity and price for statistics**

In [8]:
df[['Quantity', 'Price']].describe()

Quantity          Price
count  200000.000000  200000.000000
mean        9.670625       4.787865
std        88.196223     172.752491
min     -9600.000000  -53594.360000
25%         1.000000       1.250000
50%         3.000000       2.100000
75%        10.000000       4.150000
max     19152.000000   38970.000000

<p style="color:#00a6ff; font-size: medium"> Here we notice that there are negative values in 'Quantity' & 'Price' columns</p>

In [18]:
df['StockCode'].nunique()

4714

<p style="color:#00a6ff; font-size: medium"> We have 4714 unique stock codes means 4714 different items</p>

# Question one

<h2 style="color:#fc8803 "> Determine the total sales per invoice and identify the top 5 invoices </h2>

### Steps:
- Add column to df named total(quantity * price)
- Group data by invoice number
- Sum all totals for each invoice
- Display the top 5 invoices

In [19]:
# step 1
def calculate_total(quantity, price):
    return quantity * price

df['Total'] = calculate_total(df['Quantity'], df['Price'])
df.head()

Invoice StockCode                     Description  Quantity  \
987378  575955     22337  ANGEL DECORATION PAINTED ZINC          6   
900767  569457     22302        COFFEE MUG PEARS  DESIGN         4   
524055  538071     22767     TRIPLE PHOTO FRAME CORNICE          1   
300998  518681     21928  JUMBO BAG SCANDINAVIAN PAISLEY         4   
931091  571716     21504            SKULLS GREETING CARD        10   

                InvoiceDate  Price  Customer ID         Country  Total  
987378  2011-11-13 12:05:00   0.65      14044.0  United Kingdom   3.90  
900767  2011-10-04 11:29:00   2.55      14606.0  United Kingdom  10.20  
524055  2010-12-09 14:09:00  20.38          NaN  United Kingdom  20.38  
300998  2010-08-10 17:09:00   1.95      17799.0  United Kingdom   7.80  
931091  2011-10-18 17:05:00   0.83          NaN  United Kingdom   8.30

In [25]:
#step 2:4
top_sales_invoices = df.groupby(['Invoice'])['Total'].sum().sort_values(ascending=False)
top_sales_invoices.head()

Invoice
515944     12366.00
A563185    11062.06
576365     10226.00
572209      9066.00
493819      8852.86
Name: Total, dtype: float64

<h2 style="color:#90fc03;"> Result:</h2>

**Here we can see the top 5 sales in the selected df sample, the invoice '515944' have total of 12366.00$**

# Question 2

<h2 style="color:#fc8803 "> Compute the total sales per customer and list the top 5 customers </h2>

### Steps:
1. Group data by CustomerID
2. Select Total column
3. Sum all totals for each customer
4. Display the top 5 customers

In [26]:
top_sales_customers = df.groupby(['Customer ID'])['Total'].sum().sort_values(ascending=False)
top_sales_customers.head()

Customer ID
18102.0    134191.60
14646.0    102706.41
14156.0     57387.92
14911.0     49139.14
13694.0     39371.63
Name: Total, dtype: float64

<h2 style="color:#90fc03;"> Result:</h2>

<p>Here we can see the top 5 customers purchased in the selected df, the customer whose ID is <b>'18102.0' has a total of 134191.60$</b></p>

# Question 3

<h2 style="color:#fc8803 "> Identify the top 5 selling items during this period </h2>

### Steps:
1. Group data by StockCode
2. Select StockCode, and Quantity columns
3. Sum quantity for items
4. Display the top 5 selling items

In [27]:
items = df.groupby(['StockCode'])[['StockCode','Quantity']].sum().nlargest(5, ['Quantity'])
items

Quantity
StockCode          
37410         19414
21212         18243
85099B        17567
22197         16745
84077         14495

<h2 style="color:#90fc03;"> Result:</h2>

Here we can see the top 5 purchased items in the selected df, the items with code **'37410' has being purchased 19414 times**

# Question 4

<P style="color: #00a6ff; font-family: cursive; font-size: medium"> The retail store aims to introduce a new promotion targeting items with low sales volume. This promotion entails offering a discounted price for a pair of two different items, with one item being among the top 10 best sellers and the other from the bottom 10 sellers. Customers who purchase these two items together will qualify for the special pricing. Your task is to recommend 10 pairs of items to kickstart this campaign. </P>

<h3> Steps: </h3>
<ol>
    <li>Group data by StockCode and description, then sum up the total quantity sold for each item</li>
    <li>Save the the items where the quantity is above 0 to 'top_sold_items' var</li>
    <li>Save the the items where the quantity is equal and below 0 to 'zero_sold_items' var</li>
    <li>Create pairs, one item from the top and the other from the bottom</li>
    </ol>

In [14]:
grouped_items_sold = df.groupby(['StockCode'])['Quantity'].sum().sort_values().to_frame()
grouped_items_sold = grouped_items_sold.rename(columns={'Quantity': 'QuantitySold'})

stock_code_col = grouped_items_sold.index
grouped_items_sold['index'] = [i for i in range(len(grouped_items_sold))]
grouped_items_sold = grouped_items_sold.set_index(['index'])
grouped_items_sold['StockCode'] = stock_code_col

grouped_items_sold

QuantitySold StockCode
index                        
0            -18831     23005
1             -8142     16235
2             -6447     84347
3             -5368    72140F
4             -4552     16046
...             ...       ...
4709          14495     84077
4710          16745     22197
4711          17567    85099B
4712          18243     21212
4713          19414     37410

[4714 rows x 2 columns]

<h2 style="color: red;"> Exaplantion:</h2>

- **I grouped items based on their StockCode and summed the quantity sold for each item**
- **I converted the results to df to easy display it**
- **Finally I relpaced the index column that is generated by groupby (StockCode) with incremental index and added the StockCode column to the resulted df**

In [15]:
zero_sold_items = grouped_items_sold[grouped_items_sold['QuantitySold'] <= 0]
zero_sold_items

QuantitySold StockCode
index                        
0            -18831     23005
1             -8142     16235
2             -6447     84347
3             -5368    72140F
4             -4552     16046
...             ...       ...
308               0     90091
309               0    44091A
310               0     72732
311               0    90060D
312               0     84881

[313 rows x 2 columns]

In [16]:
top_sold_items = grouped_items_sold[grouped_items_sold['QuantitySold'] > 0].sort_values(by='QuantitySold', ascending=False)
top_sold_items

QuantitySold StockCode
index                        
4713          19414     37410
4712          18243     21212
4711          17567    85099B
4710          16745     22197
4709          14495     84077
...             ...       ...
426               1    17090D
425               1    84534b
424               1     72793
423               1     47567
313               1     90154

[4401 rows x 2 columns]

In [17]:
recommended_pairs = [
    {
        'top': top_sold_items.iloc[i]['StockCode'],
        'buttom': zero_sold_items.iloc[i]['StockCode']
    }
    for i in range(10)
]

recommended_pairs

[{'top': '37410', 'buttom': '23005'},
 {'top': '21212', 'buttom': '16235'},
 {'top': '85099B', 'buttom': '84347'},
 {'top': '22197', 'buttom': '72140F'},
 {'top': '84077', 'buttom': '16046'},
 {'top': '84879', 'buttom': '22755'},
 {'top': '85123A', 'buttom': '85160A'},
 {'top': '17003', 'buttom': '16162L'},
 {'top': '84991', 'buttom': '85184D'},
 {'top': '21977', 'buttom': '22433'}]

<h2> The list above contains a dictionary of paired items stock_code's, one item from the top sold and the other from the lowest sold </h2>